### PR TITLE
[dy] Add SLA to backend

### PR DESCRIPTION
**Pull request by dy46**
# Summary

Support checking for pipeline run SLA in backend. If the SLA passes, we will send a notification to the user if they have a notification_config set up.

# Tests

tested locally

cc: @wangxiaoyou1993

### CODE REVIEW

File: mage_ai/orchestration/pipeline_scheduler.py [hunk 1]
```diff
@@ -378,6 +381,38 @@ def run_pipeline(pipeline_run_id, variables, tags):
...
+    if pipeline_runs:
+        pipeline = Pipeline.get(pipeline_runs[0].pipeline_schedule.pipeline_uuid)
+        notification_sender = NotificationSender(
+            NotificationConfig.load(config=pipeline.repo_config.notification_config),
+        )
+
+        current_time = datetime.now(tz=pytz.UTC)
+        for pipeline_run in pipeline_runs:
+            sla = pipeline_run.pipeline_schedule.sla
+            start_date = pipeline_run.execution_date if pipeline_run.execution_date is not None else pipeline_run.created_at
```
Comment (wangxiaoyou1993) [inline mage_ai/orchestration/pipeline_scheduler.py:405]: Check whether sla is None before defining the start_date

File: mage_ai/orchestration/db/models.py [hunk 2]
```diff
@@ -130,6 +130,7 @@ class ScheduleInterval(str, enum.Enum):
     schedule_interval = Column(String(50))
     status = Column(Enum(ScheduleStatus), default=ScheduleStatus.INACTIVE)
     variables = Column(JSON)
+    sla = Column(Integer, default=None)
```
Comment (wangxiaoyou1993) [inline mage_ai/orchestration/db/models.py:133]: could you add a comment here saying this is for seconds?

File: mage_ai/orchestration/pipeline_scheduler.py [hunk 3]
```diff
@@ -378,6 +381,38 @@ def run_pipeline(pipeline_run_id, variables, tags):
...
+        ).all()
+
+    if pipeline_runs:
+        pipeline = Pipeline.get(pipeline_runs[0].pipeline_schedule.pipeline_uuid)
+        notification_sender = NotificationSender(
+            NotificationConfig.load(config=pipeline.repo_config.notification_config),
+        )
+
+        current_time = datetime.now(tz=pytz.UTC)
+        for pipeline_run in pipeline_runs:
```
Comment (wangxiaoyou1993) [inline mage_ai/orchestration/pipeline_scheduler.py:403]: Add a TODO here to combine these alerts in one notification.

File: mage_ai/orchestration/pipeline_scheduler.py [hunk 4]
```diff
@@ -378,6 +381,38 @@ def run_pipeline(pipeline_run_id, variables, tags):
...
+        PipelineRun.query.filter(
+            PipelineRun.pipeline_schedule_id.in_(pipeline_schedules),
+            PipelineRun.status.in_(IN_PROGRESS_STATUSES),
+            PipelineRun.passed_sla.is_(False),
+        ).all()
+
+    if pipeline_runs:
+        pipeline = Pipeline.get(pipeline_runs[0].pipeline_schedule.pipeline_uuid)
+        notification_sender = NotificationSender(
+            NotificationConfig.load(config=pipeline.repo_config.notification_config),
```
Comment (wangxiaoyou1993) [inline mage_ai/orchestration/pipeline_scheduler.py:399]: Or you can do `get_repo_config(get_repo_path()). notification_config`

**Review by wangxiaoyou1993 — APPROVED** [review]

File: mage_ai/orchestration/pipeline_scheduler.py [hunk 5]
```diff
@@ -378,6 +381,48 @@ def run_pipeline(pipeline_run_id, variables, tags):
...
+    pipeline_runs = \
+        PipelineRun.query.filter(
+            PipelineRun.pipeline_schedule_id.in_(pipeline_schedules),
+            PipelineRun.status.in_(IN_PROGRESS_STATUSES),
+            PipelineRun.passed_sla.is_(False),
+        ).all()
+
+    if pipeline_runs:
+        notification_sender = \
+            NotificationSender(get_repo_config(get_repo_path()).notification_config)
```
Comment (tommydangerous) [inline mage_ai/orchestration/pipeline_scheduler.py:404]: @dy46 did you check to see if this works?

cc @wangxiaoyou1993

**Pull request merged**